File: typescript/src/ships/weapons-capacitor.ts
/**
 * Weapons-capacitor recharge and endurance at a chosen WEP-pip allocation.
 *
 * A power distributor's catalogue recharge figure is its four-pip maximum. The
 * actual rate follows the same non-linear pip curve as the SYS capacitor:
 * `ratedRecharge * (weaponsPips / 4) ^ 1.1`. Endurance compares that rate with
 * the weapons' sustained draw, so magazine reloads are already folded in.
 *
 * This module is data-free. {@link ShipLoadout.weaponsCapacitorMetrics}
 * (`./ship-loadout`) resolves the powered distributor and weapons for you.
 *
 * @remarks
 * Reference implementation: EDCD/Coriolis, `src/app/shipyard/Calculations.js`
 * (`wepRechargeRate`, `timeToDrainWep`), commit
 * `68c042ca6e3db62372cbbb2077cf972345511712`. The algorithm is ported as fact,
 * not code; credit and licence terms are in
 * [ATTRIBUTIONS.md](https://github.com/DarkSession/Elite-Dangerous-Almanac/blob/main/ATTRIBUTIONS.md).
 *
 * @packageDocumentation
 */

import { capacitorRechargeAtPips } from './internal/capacitor-recharge.js';

/** Everything {@link weaponsCapacitorMetrics} needs about one firing load. */
export interface WeaponsCapacitorInput {
    /** WEP-capacitor capacity, in megajoules — finite and non-negative. */
    readonly weaponsCapacity: number;
    /** Four-WEP-pip recharge rate, in megajoules per second — finite and non-negative. */
    readonly weaponsRecharge: number;
    /** Sustained weapons-capacitor draw, in megajoules per second — finite and non-negative. */
    readonly sustainedEnergyPerSecond: number;
    /** Pips assigned to WEP, in `[0, 4]`. Fractional pips are accepted; defaults to `4`. */
    readonly weaponsPips?: number;
}

/** Recharge, drain and endurance of one weapons capacitor. */
export interface WeaponsCapacitorMetrics {
    /** Pips assigned to WEP for this result, in `[0, 4]`. */
    readonly weaponsPips: number;
    /** WEP-capacitor capacity, in megajoules. */
    readonly capacity: number;
    /** Actual recharge rate at {@link weaponsPips}, in megajoules per second. */
    readonly rechargeRate: number;
    /** Sustained draw across the firing weapons, in megajoules per second. */
    readonly sustainedEnergyPerSecond: number;
    /** Capacity lost per second after recharge, floored at `0`, in megajoules per second. */
    readonly netDrainRate: number;
    /** Seconds from full to empty, or `Infinity` when recharge keeps pace with draw. */
    readonly timeToDrain: number;
}

function requireFiniteNonNegative(scope: string, name: string, value: number): void {
    if (!Number.isFinite(value) || value < 0) {
        throw new RangeError(`${scope}: ${name} must be a finite non-negative number`);
    }
}

/**
 * Calculate WEP recharge and time to drain while a weapon load fires continuously.
 *
 * @param input - Distributor capacity, its rated four-pip recharge, sustained weapon
 * draw and the WEP-pip allocation to model.
 * @returns Actual recharge, net drain and seconds until empty. `timeToDrain` is
 * `Infinity` when the weapons draw no more than the actual recharge rate. A positive
 * draw with zero capacity drains immediately (`0` seconds).
 * @throws {RangeError} If a capacity, recharge or draw is negative or non-finite, or
 * `weaponsPips` is outside `[0, 4]`.
 * @example
 * ```ts
 * import { weaponsCapacitorMetrics } from '@elite-dangerous-almanac/core/ships/weapons-capacitor';
 *
 * weaponsCapacitorMetrics({
 *   weaponsCapacity: 20,
 *   weaponsRecharge: 5,
 *   sustainedEnergyPerSecond: 7,
 *   weaponsPips: 2,
 * }).timeToDrain; // -> 4.285…
 * ```
 */
export function weaponsCapacitorMetrics(input: WeaponsCapacitorInput): WeaponsCapacitorMetrics {
    const scope = 'weaponsCapacitorMetrics';
    const weaponsPips = input.weaponsPips ?? 4;
    if (!Number.isFinite(weaponsPips) || weaponsPips < 0 || weaponsPips > 4) {
        throw new RangeError(`${scope}: weaponsPips must be a finite number from 0 to 4`);
    }
    for (const field of [
        'weaponsCapacity',
        'weaponsRecharge',
        'sustainedEnergyPerSecond',
    ] as const) {
        requireFiniteNonNegative(scope, field, input[field]);
    }

    const rechargeRate = capacitorRechargeAtPips(input.weaponsRecharge, weaponsPips);
    const netDrainRate = Math.max(0, input.sustainedEnergyPerSecond - rechargeRate);
    return {
        weaponsPips,
        capacity: input.weaponsCapacity,
        rechargeRate,
        sustainedEnergyPerSecond: input.sustainedEnergyPerSecond,
        netDrainRate,
        timeToDrain: netDrainRate === 0 ? Infinity : input.weaponsCapacity / netDrainRate,
    };
}

File: dotnet/src/EliteDangerousAlmanac/Ships/WeaponsCapacitor.cs
using System;
using EliteDangerousAlmanac.Ships.Internal;

namespace EliteDangerousAlmanac.Ships;

/// <summary>Everything a weapons-capacitor calculation needs about one firing load.</summary>
/// <param name="WeaponsCapacity">The WEP-capacitor capacity, in megajoules.</param>
/// <param name="WeaponsRecharge">The four-WEP-pip recharge rate, in megajoules per second.</param>
/// <param name="SustainedEnergyPerSecond">
/// The sustained weapons-capacitor draw, in megajoules per second.
/// </param>
/// <remarks>Every figure must be finite and zero or more.</remarks>
public sealed record WeaponsCapacitorInput(
    double WeaponsCapacity,
    double WeaponsRecharge,
    double SustainedEnergyPerSecond);

/// <summary>The recharge, the drain and the endurance of one weapons capacitor.</summary>
/// <param name="WeaponsPips">The pips assigned to WEP for this result, from zero through four.</param>
/// <param name="Capacity">The WEP-capacitor capacity, in megajoules.</param>
/// <param name="RechargeRate">The recharge rate at the allocation, in megajoules per second.</param>
/// <param name="SustainedEnergyPerSecond">
/// The sustained draw across the firing weapons, in megajoules per second.
/// </param>
/// <param name="NetDrainRate">
/// The capacity lost per second after the recharge, in megajoules per second, and never below
/// nothing.
/// </param>
/// <param name="TimeToDrain">
/// The seconds from full to empty, or infinity where the recharge keeps pace with the draw.
/// </param>
public sealed record WeaponsCapacitorMetrics(
    double WeaponsPips,
    double Capacity,
    double RechargeRate,
    double SustainedEnergyPerSecond,
    double NetDrainRate,
    double TimeToDrain);

/// <summary>The weapons capacitor's recharge and endurance at a chosen WEP-pip allocation.</summary>
/// <remarks>
/// <para>
/// A distributor's catalogue recharge figure is its four-pip maximum, and the actual rate follows
/// the same non-linear pip curve as the SYS capacitor. The endurance compares that rate with the
/// weapons' sustained draw, so the magazine reloads are already folded in.
/// </para>
/// <para>
/// Reference implementation: EDCD/Coriolis, <c>src/app/shipyard/Calculations.js</c>. The algorithm
/// is ported as fact rather than as code. See <c>ATTRIBUTIONS.md</c> for credit and licence terms.
/// </para>
/// </remarks>
public static class WeaponsCapacitor
{
    /// <summary>
    /// Calculates the WEP recharge and the time to drain while a weapon load fires continuously.
    /// </summary>
    /// <param name="input">
    /// The distributor capacity, its rated four-pip recharge and the weapons' sustained draw.
    /// </param>
    /// <param name="weaponsPips">
    /// The pips assigned to WEP, from zero through four. Fractional pips are accepted.
    /// </param>
    /// <returns>
    /// The actual recharge, the net drain and the seconds until empty. The time to drain is
    /// infinite where the weapons draw no more than the actual recharge rate, and a draw against
    /// no capacity at all drains in no time.
    /// </returns>
    /// <exception cref="ArgumentNullException"><paramref name="input"/> is <see langword="null"/>.</exception>
    /// <exception cref="ArgumentOutOfRangeException">
    /// A capacity, recharge rate or draw is negative or not finite, or the allocation is outside
    /// zero through four.
    /// </exception>
    public static WeaponsCapacitorMetrics Metrics(
        WeaponsCapacitorInput input, double weaponsPips = 4)
    {
        RangeGuards.RequirePips(weaponsPips, nameof(weaponsPips));
        if (input is null) throw new ArgumentNullException(nameof(input));

        RangeGuards.RequireFiniteNonNegative(input.WeaponsCapacity, "input.WeaponsCapacity");
        RangeGuards.RequireFiniteNonNegative(input.WeaponsRecharge, "input.WeaponsRecharge");
        RangeGuards.RequireFiniteNonNegative(
            input.SustainedEnergyPerSecond, "input.SustainedEnergyPerSecond");

        double rechargeRate = CapacitorRecharge.AtPips(input.WeaponsRecharge, weaponsPips);
        double netDrainRate = Math.Max(0, input.SustainedEnergyPerSecond - rechargeRate);

        return new WeaponsCapacitorMetrics(
            WeaponsPips: weaponsPips,
            Capacity: input.WeaponsCapacity,
            RechargeRate: rechargeRate,
            SustainedEnergyPerSecond: input.SustainedEnergyPerSecond,
            NetDrainRate: netDrainRate,
            TimeToDrain: netDrainRate == 0
                ? double.PositiveInfinity
                : input.WeaponsCapacity / netDrainRate);
    }
}
